validation/validation_test.go

```diff
@@ -17,54 +22,92 @@ var (
 	runtime = "runc"
 )
 
+type validation struct {
+	test        func(string, string, *rspec.Spec) error
+	description string
+}
+
 func init() {
 	runtime = os.Getenv("RUNTIME")
 }
 
-func runtimeValidate(runtime string, g *generate.Generator) error {
+func TestValidateRuntimeInside(t *testing.T) {
+	g, err := getDefaultGenerator()
+	if err != nil {
+		t.Errorf("%s failed validation: %v", runtime, err)
+	}
+	g.SetProcessArgs([]string{"/runtimetest"})
+
+	if err := runtimeInsideValidate(runtime, g); err != nil {
+		t.Errorf("%s failed validation: %v", runtime, err)
+	}
+}
+
+func TestValidateRuntimeOutside(t *testing.T) {
+	g, err := getDefaultGenerator()
+	if err != nil {
+		t.Errorf("%s failed validation: %v", runtime, err)
+	}
+
+	if err := runtimeOutsideValidate(runtime, g); err != nil {
+		t.Errorf("%s failed validation: %v", runtime, err)
+	}
+}
+
+func runtimeInsideValidate(runtime string, g *generate.Generator) error {
 	// Find the runtime binary in the PATH
 	runtimePath, err := exec.LookPath(runtime)
 	if err != nil {
 		return err
 	}
 
-	// Setup a temporary test directory
-	tmpDir, err := ioutil.TempDir("", "ocitest")
+	bundleDir, rootfsDir, err := prepareBundle(g)
 	if err != nil {
 		return err
 	}
-	defer os.RemoveAll(tmpDir)
+	//	defer os.RemoveAll(bundleDir)
 
-	// Create bundle directory for the test container
-	bundleDir := tmpDir + "/busybox"
-	if err := os.MkdirAll(bundleDir, 0755); err != nil {
+	// Copy the runtimetest binary to the rootfs
+	err = fileutils.CopyFile("../runtimetest", filepath.Join(rootfsDir, "runtimetest"))
```

Comment:
Does this actually work?  My local build of `runtimetest` is dynamically linked:

```
$ lddtree runtimetest 
runtimetest (interpreter => /lib64/ld-linux-x86-64.so.2)
    libpthread.so.0 => /lib64/libpthread.so.0
    libc.so.6 => /lib64/libc.so.6
```

and neither that linker nor those libraries aren't part of our Busybox `rootfs`.  For example, here is the statically-linked Busybox working fine in a chroot based on that `rootfs`:

```
$ unshare -rUmfp --mount-proc sh -c 'mount --rbind /proc proc && chroot . sh -c "pwd && ls"'
/
bin          etc          proc         runtimetest
```

And here is `runtimetest` failing to launch in the same situation:

```
$ unshare -rUmfp --mount-proc sh -c 'mount --rbind /proc proc && chroot . /runtimetest'
chroot: failed to run command ‘/runtimetest’: No such file or directory
```

I expect we're going to need to either compile `runtimetest` as a static binary (#442), copy the linker and required libraries into `rootfs`, or use `busybox` for our internal tests.